File: services/web/apps/fm/alarm/plugins/validation.py
# -*- coding: utf-8 -*-
# ---------------------------------------------------------------------
# ValidationPlugin
# ---------------------------------------------------------------------
# Copyright (C) 2007-2015 The NOC Project
# See LICENSE for details
# ---------------------------------------------------------------------

# Django modules
from django.template import Template, Context
from noc.cm.models.errortype import ErrorType
from noc.cm.models.objectfact import ObjectFact

# NOC modules
from base import AlarmPlugin


class ValidationPlugin(AlarmPlugin):
    name = "validation"

    def get_data(self, alarm, config):
        r = {
            "plugins": [("NOC.fm.alarm.plugins.Validation", {})],
            "validation_errors": []
        }
        t_cache = {}
        for f in ObjectFact.objects.filter(
                object=alarm.managed_object.id,
                cls="error").order_by("introduced"):
            etn = f.attrs.get("type")
            if etn:
                et = t_cache.get(etn)
                if not et:
                    et = ErrorType.objects.filter(name=etn).first()
                    t_cache[etn] = et
                subject = Template(et.subject_template).render(Context(f.attrs))
                body = Template(et.body_template).render(Context(f.attrs))
            else:
                subject = etn
                body = etn

            r["validation_errors"] += [{
                "uuid": str(f.uuid),
                "introduced": f.introduced.isoformat(),
                "changed": f.changed.isoformat(),
                "subject": subject,
                "body": body,
                "cls": etn,
                "attrs": f.attrs
            }]
        return r

File: ui/web/fm/alarm/plugins/Validation.js
//---------------------------------------------------------------------
// Validation plugin
//---------------------------------------------------------------------
// Copyright (C) 2007-2015 The NOC Project
// See LICENSE for details
//---------------------------------------------------------------------
console.debug("Defining NOC.fm.alarm.plugins.Validation");

Ext.define("NOC.fm.alarm.plugins.Validation", {
  extend: "Ext.panel.Panel",
  title: __("Validation"),
  app: null,
  autoScroll: true,
  bodyPadding: 4,
  layout: "border",
  border: false,
  detailTpl: Ext.create("Ext.Template", [
    "<div style='border-bottom: 1px solid gray'>",
    "<b>Subject:</b> {subject}",
    "</div>",
    "<pre>{body}</pre>",
  ]),

  initComponent: function(){
    var me = this;

    me.store = Ext.create("Ext.data.Store", {
      fields: [
        "uuid", "subject", "body",
        {
          name: "introduced",
          type: "date",
        },
        {
          name: "changed",
          type: "date",
        },
        "cls",
      ],
      data: [],
    });

    me.selectGrid = Ext.create("Ext.grid.Panel", {
      //autoScroll: true,
      store: me.store,
      columns: [
        {
          dataIndex: "subject",
          text: __("Subject"),
          flex: 1,
        },
        {
          dataIndex: "cls",
          text: __("Error Type"),
          width: 200,
        },
        {
          dataIndex: "introduced",
          text: __("Introduced"),
          width: 150,
          renderer: NOC.render.DateTime,
        },
        {
          dataIndex: "changed",
          text: __("Changed"),
          width: 150,
          renderer: NOC.render.DateTime,
        },
      ],
      height: 200,
      forceFit: true,
      region: "north",
      listeners: {
        scope: me,
        selectionchange: me.onSelectionChange,
      },
    });

    me.detailPanel = Ext.create("Ext.panel.Panel", {
      region: "center",
      bodyPadding: 8,
      html: "Please select an error to see details",
    });

    Ext.apply(me, {
      items: [
        me.selectGrid,
        me.detailPanel,
      ],
    });

    me.callParent();
  },

  updateData: function(data){
    var me = this;
    me.store.loadData(data.validation_errors);
  },

  onSelectionChange: function(sm, record){
    var me = this;
    if(record.length){
      me.detailPanel.update(
        me.detailTpl.apply(record[0].data),
      );
    }
  },
});


